### PR TITLE
Add user following

### PR DESCRIPTION
**Pull request by thanhconguet214**

### CODE REVIEW

**Comment by thanhconguet214** [comment]
ready

File: app/controllers/relationships_controller.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class RelationshipsController < ApplicationController
+  before_action :logged_in_user
+  def create
+    @user = User.find_by id: params[:followed_id]
+    current_user.follow @user
+    respond_to do |format|
+      format.js
+    end
+  end
+  def destroy
```
Comment (mdmahmudurrahman) [inline app/controllers/relationships_controller.rb]: Add a new line between 2 actions

File: app/models/user.rb [hunk 2]
```diff
@@ -1,5 +1,13 @@
   class User < ApplicationRecord
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  has_secure_password
+  has_many :microposts, dependent: :destroy
+  has_many :active_relationships, class_name:  "Relationship",
+    foreign_key: "follower_id", dependent: :destroy
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: `class_name:  Relationship.name, foreign_key: :follower_id`

File: app/models/user.rb [hunk 3]
```diff
@@ -1,5 +1,13 @@
...
+  has_many :passive_relationships, class_name:  "Relationship",
+    foreign_key: "followed_id", dependent: :destroy
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: The same here

File: app/views/relationships/create.js.erb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+$('#follow_form').html("<%= j render'users/unfollow' %>");
```
Comment (mdmahmudurrahman) [inline app/views/relationships/create.js.erb]: 1. Use single quote in CSS
2. Double quote in ruby
In summary
`$('#follow_form').html('<%= j render "users/unfollow" %>')`

File: app/views/relationships/create.js.erb [hunk 5]
```diff
@@ -0,0 +1,2 @@
...
+$('#followers').html("<%= @user.followers.count %>");
```
Comment (mdmahmudurrahman) [inline app/views/relationships/create.js.erb]: Remove redundant ()

File: app/views/relationships/destroy.js.erb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+$('#follow_form').html("<%= j render 'users/follow' %>");
+$('#followers').html("<%= @user.followers.count %>");
```
Comment (mdmahmudurrahman) [inline app/views/relationships/destroy.js.erb]: The same here

File: app/views/shared/_stats.html.erb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+<% @user ||= current_user %>
+<div class="stats">
+  <a href="<%= following_user_path(@user) %>">
```
Comment (mdmahmudurrahman) [inline app/views/shared/_stats.html.erb]: Remove ()

File: app/views/shared/_stats.html.erb [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+    <strong id="following" class="stat">
+      <%= @user.following.count %>
+    </strong>
+    <%= t "shared._stats.following" %>
+  </a>
+  <a href="<%= followers_user_path(@user) %>">
```
Comment (mdmahmudurrahman) [inline app/views/shared/_stats.html.erb]: The same here

File: app/views/users/_unfollow.html.erb [hunk 9]
```diff
@@ -0,0 +1,4 @@
+<%= form_for(current_user.active_relationships.find_by(followed_id: @user.id),
+      html: {method: :delete}) do |f| %>
```
Comment (mdmahmudurrahman) [inline app/views/users/_unfollow.html.erb]: Indentation

File: app/views/users/_unfollow.html.erb [hunk 10]
```diff
@@ -0,0 +1,4 @@
+<%= form_for(current_user.active_relationships.find_by(followed_id: @user.id),
```
Comment (mdmahmudurrahman) [inline app/views/users/_unfollow.html.erb]: Use helper method for building the form for this 
`current_user.active_relationships.find_by(followed_id: @user.id)`

File: app/views/users/show_follow.html.erb [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+      <% end %>
+    </section>
+  </aside>
+  <div class="col-md-8">
+    <h3><%= @title %></h3>
+    <% if @users.any? %>
+      <ul class="users follow">
+        <%= render @users %>
+      </ul>
+      <%= paginate @users%>
```
Comment (mdmahmudurrahman) [inline app/views/users/show_follow.html.erb]: `<%= paginate @users %>`

File: db/seeds.rb [hunk 12]
```diff
@@ -22,4 +22,11 @@
 50.times do
   content = FFaker::Lorem.sentence(5)
```
Comment (mdmahmudurrahman) [inline db/seeds.rb]: Remove ()

File: db/seeds.rb [hunk 13]
```diff
@@ -22,4 +22,11 @@
...
   users.each {|user| user.microposts.create!(content: content)}
-end
+end 
+
+users = User.all
+user  = users.first
+following = users[2..50]
+followers = users[3..40]
+following.each {|followed| user.follow(followed)}
+followers.each {|follower| follower.follow(user)}
```
Comment (mdmahmudurrahman) [inline db/seeds.rb]: Remove ()

File: test/controllers/relationships_controller_test.rb [hunk 14]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (mdmahmudurrahman) [inline test/controllers/relationships_controller_test.rb]: Remove this file

File: test/fixtures/relationships.yml [hunk 15]
```diff
@@ -0,0 +1,9 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (mdmahmudurrahman) [inline test/fixtures/relationships.yml]: The same here

File: test/models/relationship_test.rb [hunk 16]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (mdmahmudurrahman) [inline test/models/relationship_test.rb]: The same here

**Comment by mdmahmudurrahman** [comment]
commented

**Comment by thanhconguet214** [comment]
ready

**Comment by huuphuc93** [comment]
commented

**Comment by thanhconguet214** [comment]
ready

File: app/views/users/_unfollow.html.erb [hunk 17]
```diff
@@ -0,0 +1,3 @@
+<%= form_for unfollow_user, html: method: :delete do |f| %>
```
Comment (huuphuc93) [inline app/views/users/_unfollow.html.erb]: html: method: :delete => method: :delete

**Comment by huuphuc93** [comment]
commented
comment nhỏ

File: app/helpers/relationships_helper.rb [hunk 18]
```diff
@@ -0,0 +1,9 @@
+module RelationshipsHelper
+  def unfollow_user
+    current_user.active_relationships.find_by(followed_id: @user.id)
+  end
+
+  def follow_user
+  	current_user.active_relationships.build
```
Comment (huuphuc93) [inline app/helpers/relationships_helper.rb]: indent

**Comment by thanhconguet214** [comment]
ready

File: app/controllers/relationships_controller.rb [hunk 19]
```diff
@@ -0,0 +1,21 @@
+class RelationshipsController < ApplicationController
+  before_action :logged_in_user
+
+  def create
+    @user = User.find_by id: params[:followed_id]
+    current_user.follow @user
+    respond_to do |format|
+      format.html {redirect_to @user}
```
Comment (huuphuc93) [inline app/controllers/relationships_controller.rb]: format.html {redirect_to @user} có cần dùng k? Bỏ đi nó có chạy k?

File: app/helpers/relationships_helper.rb [hunk 20]
```diff
@@ -0,0 +1,9 @@
+module RelationshipsHelper
+  def unfollow_user
+    current_user.active_relationships.find_by(followed_id: @user.id)
```
Comment (huuphuc93) [inline app/helpers/relationships_helper.rb]: remove () dc k nhỉ?

**Comment by huuphuc93** [comment]
commented
comment nhỏ

File: app/controllers/relationships_controller.rb (hunk 19 above)
Comment (thanhconguet214) [reply app/controllers/relationships_controller.rb]: bỏ đi thì unfollow của em bị lỗi

**Comment by thanhconguet214** [comment]
ready

**Comment by huuphuc93** [comment]
merged
this (pr) can be (merged) now
CC: [To:1818111] Nguyen Thi Huong (94)
CC: [To:2076564] Nguyen Huy Hung C (93)
CC: [To:1546746] Md. Mahmudur Rahman (Mahmud)

**Pull request merged**